### DISCUSSION TITLE
Making pyright aware of pytest-mock types

### DISCUSSION THREAD

**Discussion by gregspurrier** [microsoft/pyright#5311] [Q&A] [answered]
I'm new to `pyright` and I'm trying to get my `pytest` tests to type check in strict mode. I'm running into troubles with fixtures defined by `pytest-mock`. 

My `pyrightconfig.json` is very simple:

```json
{
    "typeCheckingMode": "strict"
}
```

With this test:

```python
def test_something(mocker):
    foo = mocker.Mock()
    foo.return_value = 37
```

I'm seeing the following output from `pytest` (paths elided for brevity):

```
.../tests/test_foo.py
  .../tests/test_foo.py:1:20 - error: Type of parameter "mocker" is unknown (reportUnknownParameterType)
  .../tests/test_foo.py:1:20 - error: Type annotation is missing for parameter "mocker" (reportMissingParameterType)
  .../tests/test_foo.py:2:11 - error: Type of "Mock" is unknown (reportUnknownMemberType)
  .../tests/test_foo.py:2:5 - error: Type of "foo" is unknown (reportUnknownVariableType)
4 errors, 0 warnings, 0 informations 
```

My understanding is that `pytest-mock` has type information. When I hover over the `mocker` parameter in VS Code, it shows its type, so the type information does appear to exist and be available in my environment. I just don't know what I need to do in order for `pyright` to find it when invoked from the command line.

Any pointers would be appreciated.

Versions:
* Python 3.10.10
* pytest 6.2.5
* pytest-mock 3.6.1
* pyright 1.1.314

**Comment by erictraut** [comment, marked as answer]
If you want to use strict-mode rules, you need to provide type annotations for all of your function input parameters. In your example above, `mocker` is unannotated.

The reason you're seeing type information here is because pylance (the language server built on top of pyright) has some limited knowledge of pytest fixtures. Pyright, the underlying type checker, does not. Pyright is a standards-based type checker, and it has no knowledge of specific third-party libraries like pytest. All of its type information comes from type stubs, type annotations within the code, or information it can infer directly from the code. You may find this "[getting started](https://microsoft.github.io/pyright/#/type-concepts)" documentation useful if you're new to static type checking.

I also usually recommend that developers who are new to static type checking start with "basic" type checking rules, not "strict". The strict-mode rules produce more errors (some of which could be false positives), and it requires more advanced knowledge to address these errors.

**Reply by gregspurrier** [reply]
Oh yes, of course. I'd completely missed that I'd not given a type for the parameter and was fuzzy about what pylance offers vs what pyright does. Thank you for your patience and help. 

I think what I might want to do is use strict for production code and basic for tests. Is that a bad idea?

**Reply by erictraut** [reply]
> I think what I might want to do is use strict for production code and basic for tests. Is that a bad idea?

That's a fine approach — one that I've used in a number of my projects. 

You can specify individual directories within your project should use strict-mode settings using the ["strict" config option](https://microsoft.github.io/pyright/#/configuration?id=main-configuration-options). All other files will default to "basic" (or whatever other diagnostic rules you've overridden manually in your config). The "strict" config option accepts globs specifiers, so you can even use file or directory name patterns to specify which files should be strict. 

Using this same approach, you can start with "basic" type checking for your entire project but then add strict-mode checking on a file-by-file or directory-by-directory basis as you make improvements.

**Reply by gregspurrier** [reply]
That makes sense and I'll go with that approach. Thanks again for your help!
